src/mlpack/methods/ann/vanilla_rbm.hpp

```diff
@@ -0,0 +1,208 @@
+  * 
+  */
+  double Evaluate(const arma::mat& parameters, arma::vec& orderFunction);
+
+ /** 
+  * This function calculates
+  * the free energy of the model
+  * @param: input data point 
+  */
+  double FreeEnergy(const arma::mat&& input);
+
+ /*
+  * This functions samples the hidden
+  * layer given the visible layer
+  *
+  * @param input visible layer input
+  * @param output the sampled hidden layer
+  */
+  void SampleHidden(arma::mat&& input, arma::mat&& output);
+
+  /*
+  * This functions samples the visble
+  * layer given the hidden layer
+  *
+  * @param input hidden layer
+  * @param output the sampled visible layer 
+  */
+  void SampleVisible(arma::mat&& input, arma::mat&& output);
+
+ /*
+  * This function does the k-step
+  * gibbs sampling.
+  *
+  * @param negative_sample: stores the negative sample
+  * @param num_step number of steps in gibbs sampling
+  * @param persistnce start chain at input or the previous negative_sample
+  */
+  void Gibbs(arma::mat&& input, 
+             arma::mat&& negative_sample, 
+             size_t num_steps = 1, 
+             bool persistence = false);
+
+
+  /*
+   * Calculates the gradients for the rbm network
+   *
+   * @param input index the visible layer/data point
+   * @param neg_input stores the negative samples computed usign the gibbs 
+   * @param k number of steps for gibbs sampling
+   * @param persistence pcdk / not 
+   * @param output store the gradients
+   */
+  void Gradient(const size_t input, 
+                const size_t k, 
+                const bool persistence, 
+                arma::mat& output);
+
+  //! Return the number of separable functions (the number of predictor points).
+  size_t NumFunctions() const { return numFunctions; }
+
+  //! Return the initial point for the optimization.
+  const arma::mat& Parameters() const { return parameter; };
+  //! Modify the initial point for the optimization.
+  arma::mat& Parameters() { return parameter; };
+  
+  //! Serialize the model.
+  template<typename Archive>
+  void Serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+
+  /* 
+   * ForwardVisible layer compute the forward
+   * activations given the visible layer
+   *
+   * @param input the visible layer
+   * @param output the acitvation function
+   */
+  void ForwardVisible(arma::mat&& input, arma::mat&& output)
+  {
+    visible.Forward(std::move(input), std::move(output));
+  };
+
+  /* 
+   * ForwardHidden layer compute the forward
+   * activations given the hidden layer
+   *
+   * @param input the visible layer
+   * @param output the acitvation function
+   */
+  void ForwardHidden(arma::mat&& input, arma::mat&& output)
+  {
+    hidden.Forward(std::move(input), std::move(output));
+  };
+
+  /*
+   * Helper function for Gradient
+   * calculates the gradients for both
+   * positive and negative samples.
+   */
+  void CalcGradient(arma::mat&& input, arma::mat&& output)
+  {
+    arma::mat temp1, temp2;
```

Comment:
Could you reduce memory allocations here? As for me it is better to define these temporary variables as class members.